scipy/signal/_signaltools.py

```diff
@@ -2452,6 +2458,271 @@ def hilbert2(x, N=None):
+    the signal. Hence, `|a(t)|` "envelopes" the class of all signals with amplitude
+    `a(t)` and arbitrary phase `phi(t)`.
+    For real-valued signals, `x(t) = a(t) * cos(phi(t))` is the analogous formulation.
+    `|a(t)|` can be determined by converting `x(t)` into an analytic signal `z_a(t)`
+    by means of a Hilbert transform, i.e.,
+    `z_a(t) = a(t) * cos(phi(t)) + 1j * a(t) * sin(phi(t))`, which produces a
+    complex-valued signal with the same envelope `|a(t)|`.
+
+    Parameters
+    ----------
+    z :
+        Real- or complex-valued input signal, which is assumed to be made up of `n`
+        samples and having sampling interval `T`. `z` may also be a multidimensional
+        signal with the time axis being defined by `axis`.
+    bp_in :
+        2-tuple defining the frequency band ``bp_in[0]:bp_in[1]`` of the input filter.
+        The corner frequencies are specified as multiples of `1/(n*T)` and the allowed
+        frequency range is specified by ``-n//2 <= bp_in[0] < bp_in[1] <= (n+1)//2``.
+        ``None`` entries are replaced with `-n//2` or `(n+1)//2` respectively. The
+        default of ``(1, None)`` removes the mean value as well as the negative
+        frequency components.
+    n_out :
+        If not ``None`` the output will be resampled to `n_out` samples.
+    squared :
+        If set, the square of the envelope is returned. This is useful, if filters need
+        to be applied to the result, since the squared envelope possesses fewer
+        harmonics.
+    residual :
+        This option determines what kind of residual, i.e., the signal part which the
+        input bandpass filter removes, is returned. `'all'` returns everything except
+        the contents of the frequency band ``bp_in[0]:bp_in[1]``, `'lowpass'`
+        returns the contents of the frequency band ``< bp_in[0]``. If ``None`` then
+        nothing is returned.
+    axis :
+       Axis of `z` over which to compute the envelope (default is last axis).
+
+    Returns
+    -------
+    z_env :
+        The envelope of input `x`.
+    x_res :
+        Optional residual, i.e., the signal part which the input bandpass filter
+        removed. What is returned is determined by the parameter `residual`. Note that
+        for real-valued signals, also a real-valued residual is returned. Hence, the
+        negative frequency components of `bp_in` are ignored.
+
+    Notes
+    -----
+    The implementation is based on computing the FFT of the input signal and then
+    performing the necessary operations in Fourier space. Hence, the typical FFT
+    caveats need to be taken into account:
+
+    * The signal is assumed to be periodic. Discontinuities between signal start and
+      end can lead to unwanted results due to Gibbs phenomenon.
+    * The FFT is slow if the signal length is prime or very long. Also, the memory
+      demands are typically higher than a comparable FIR/IIR filter based
+      implementation.
```

Comment:
Is there a way to make this implementation robust to large prime numbers? Large prime signal length occurs relatively often.

Reply:
AFAIK, not really. Depending on circumstances, either resampling or zero-padding  (consult [next_fast_len()]) might produce better results. Furthermore, [Lyons] (listed in the reference), provides further ways to calculate the envelope for different kinds of signals. Which approach is best heavily depends on the application.



[next_fast_len()]: https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.next_fast_len.html
[Lyons]: <https://community.infineon.com/gfawx74859/attachments/gfawx74859/psoc135/46469/1/R.%20Lyons_envelope_detection_v3.pdf>